### PR TITLE
[#1291] Util methods for Fides configuration scripts

### LINKED ISSUE #1291: Create initial set of "test setup scripts" to configure local Fides environment with basic data
### Description

Create a default set of “test setup scripts” that configure a Fides instance with enough data to exercise all core features, including:

* custom Fides resources: taxonomies, systems, datasets, policies
* Admin UI users
* DSR connectors: database, SaaS, manual
* DSR policies (privacy request rules)
* privacy requests (in all statuses)
* storage & email configs
* etc.

Note that this will also need a clean, standard way to inject secrets for connectors, user passwords, 3rd party test systems, etc.

We've built several versions of these _kinds_ of scripts, but have never invested enough time & energy to making them maintainable. One of the main reasons is that we've not built automated tests that rely on the scripts enough to force them to keep working and stay up to date. For some examples, see:
* the `fidesdemo` repo: https://github.com/ethyca/fidesdemo/blob/main/flaskr/fidesops.py
* the `fidesops` quickstart: https://github.com/ethyca/fides/blob/main/scripts/quickstart.py
* the `seed_test_data` command: https://github.com/ethyca/fides/blob/main/scripts/create_test_data.py
* etc.

These scripts will be used for at least three purposes over time:
1. Setting up a local environment with test data for engineers & product managers to test features
2. Setting up a staging environment with test data for automated CI checks to test features
3. Setting up a hosted environment with test data for giving product & sales demos

Because of this, it's important to make these scripts *as portable as possible*; ideally, these are relatively vanilla Python scripts that just leverage Fides CLI & API commands without calling source code functions at all.

#### Test Setup Scripts Framework
* MUST save setup scripts in a directory structure that is easy to intuit and maintain, e.g. `scripts/setup/systems.py`, `scripts/setup/postgres_connector.py`, or similar
* MUST have a `nox -s setup_test` (or similar name!) that runs all "basic" scripts to fully configure the local Fides database
* MUST NOT rely on `pytest` features or functionality; these scripts are intended to be reused in the future for E2E runners (e.g. Cypress) and deployed environments without pytest
* MUST inject secrets using ENV variables for reusability (NOTE: within `nox`, these ENV vars could be auto-populated by Vault)

#### Basic Test Setup Scripts 

To get our local environment populated with enough "basic" data to test our core features, we MUST have scripts that:
1. configure a "Fides API client" with OAuth token to test protected APIs
1. configure Users to test the Admin UI
1. configure Systems to test the data map
1. configure Datasets to test the data map
1. configure custom Data Categories, Data Uses, and Data Subjects to test the data map
1. configure an S3 Storage location to test privacy requests
1. configure a Mailgun email to test privacy requests
1. configure identity verification for the privacy center to test privacy requests
1. configure DSR access & erasure policies to test privacy requests
1. configure a Postgres connector (and dataset) to test privacy requests (using existing integration test scripts for Postgres)
1. configure a Mongodb connector (and dataset) to test privacy requests (using existing integration test scripts for Mongodb)
1. configure a Mailchimp connector (and dataset) to test privacy requests (using existing integration test scripts for Mailchimp)
1. configure a Stripe connector (and dataset) to test privacy requests (using existing integration test scripts for Stripe)
1. configure privacy requests in various statuses

#### Manual Testing
Once all basic setup scripts have run, a tester should be able to:
1. Login to the Admin UI as a User and view existing Users in User Management UI
3. View existing Systems, Datasets, and custom Taxonomy resources in their respective screens
4. View existing Privacy Requests and Connectors in their respective screens
5. Submit and execute a new privacy request via the Privacy Center & Admin UI

### PR DESCRIPTION
**Pull request by seanpreston**
Closes #1291 

### Requirements

- [x] MUST save setup scripts in a directory structure that is easy to intuit and maintain, e.g. scripts/setup/systems.py, scripts/setup/postgres_connector.py, or similar
- [x] MUST have a `nox -s setup_test` (or similar name!) that runs all "basic" scripts to fully configure the local Fides database
- [x] MUST NOT rely on pytest features or functionality; these scripts are intended to be reused in the future for E2E runners (e.g. Cypress) and deployed environments without pytest
- [x] MUST inject secrets using ENV variables for reusability (NOTE: within nox, these ENV vars could be auto-populated by Vault)
- [x] configure a "Fides API client" with OAuth token to test protected APIs
- [x] configure Users to test the Admin UI
- [x] configure an S3 Storage location to test privacy requests
- [x] configure a Mailgun email to test privacy requests
- [x] configure identity verification for the privacy center to test privacy requests
- [x] configure DSR access & erasure policies to test privacy requests
- [x] configure a Postgres connector (and dataset) to test privacy requests (using existing integration test scripts for Postgres)
- [x] configure a Mongodb connector (and dataset) to test privacy requests (using existing integration test scripts for Mongodb)
- [x] configure a Mailchimp connector (and dataset) to test privacy requests (using existing integration test scripts for Mailchimp)
- [x] configure a Stripe connector (and dataset) to test privacy requests (using existing integration test scripts for Stripe)
- [x] configure privacy requests in various statuses

### Pre-Merge Checklist

* [x] All CI Pipelines Succeeded
* Documentation Updated:
  * [ ] documentation complete, or draft/outline provided (tag docs-team to complete/review on this branch)
  * [ ] documentation issue created (tag docs-team to complete issue separately)
* [x] Issue Requirements are Met
* [x] Relevant Follow-Up Issues Created
* [ ] Update `CHANGELOG.md`

### Description Of Changes

- Included is a collection of files containing utility methods for configuring parts of Fides against a running server.
- The required secrets can be found [here](https://start.1password.com/open/i?a=5LPILPRDUBGWBPJWYS6QHHHLRE&v=42yuohsgd3pqfbepwcth7uigue&i=wjk46p25hvcfhaylrb7gycirm4&h=ethyca.1password.com)

### Follow-ups

- I've moved the `System`, `Dataset` et al scripts to https://github.com/ethyca/fides/issues/1417

### CODE REVIEW

File: scripts/example_script.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+print("Running an example Fides configuration script...")
+
+try:
+    check_health()
+except RuntimeError:
+    print(
+        f"Connection error. Please ensure Fides is healthy and running at {constants.FIDESOPS_URL}."
+    )
+
+auth_header = get_auth_header()
```
Comment (seanpreston) [inline scripts/example_script.py:19]: I'm not a huge fan of this pattern of passing the `auth_header` into every function. It makes more sense than having the script fetch a new one each time, and absent any class structure to attach the header to this appears to be the next best thing.

File: scripts/setup/authentication.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+logger = logging.getLogger(__name__)
+logging.basicConfig(level=logging.INFO)
+
+
+def get_access_token(
+    client_id: str,
+    client_secret: str,
+) -> str:
+    """
+    Authorize with fidesops via OAuth.
```
Comment (ThomasLaPiana) [inline scripts/setup/authentication.py:20]: I don't think we're delineating between ctl/ops oauth anymore, they all use the same user endpoints/authentication

File: scripts/setup/policy.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    response = requests.patch(
+        f"{constants.BASE_URL}{urls.POLICY_LIST}",
+        headers=auth_header,
+        json=policy_create_data,
+    )
+
+    if response.ok:
+        policies = (response.json())["succeeded"]
+        if len(policies) > 0:
+            logger.info("Created fidesops policy with key=%s via /api/v1/policy", key)
```
Comment (ThomasLaPiana) [inline scripts/setup/policy.py:35]:
```suggestion
            logger.info("Created fidesops policy with key=%s via /api/v1/dsr/policy", key)
```

File: scripts/setup/constants.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
+from fides.api.ops.api.v1 import urn_registry as urls
+
+FIDESOPS_URL = "http://localhost:8080"
```
Comment (ThomasLaPiana) [inline scripts/setup/constants.py:3]:
```suggestion
FIDES_URL = "http://localhost:8080"
```

File: scripts/example_script.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+from setup.user import create_user
+
+
+print("Running an example Fides configuration script...")
+
+try:
+    check_health()
+except RuntimeError:
+    print(
+        f"Connection error. Please ensure Fides is healthy and running at {constants.FIDESOPS_URL}."
```
Comment (ThomasLaPiana) [inline scripts/example_script.py:16]:
```suggestion
        f"Connection error. Please ensure Fides is healthy and running at {constants.FIDES_URL}."
```

**Comment by ThomasLaPiana** [comment]
@seanpreston for the full test environment, we need to populate _every_ possible connector/feature for testing. Maybe we can leverage some of the test setup we have to make this more complete?

**Comment by NevilleS** [comment]
Hey @ThomasLaPiana and @seanpreston - I don't think we need to populate _every_ connector, at least in this first pass. I'd like to tightly scope this and get this merged quickly, so that we can build tests on top of this base. We can definitely consider an "advanced" setup script & command (e.g. `nox -s manual_test(all databases)`) that configured _everything_ but I'd recommend we ship this first!

I'd also be fine to cut the "configure privacy requests in various statuses" requirement too, since we don't currently support some kind of "import privacy request" API where you could preload historical requests that were completed previously, etc. Therefore we'd have to "fake" data to make it work, which is a different kind of test setup than the rest.

**Comment by ThomasLaPiana** [comment]
@NevilleS can we tightly declare what the requirements here are then? Is it:

* database, webserver, worker, redis, admin UI, privacy center, example postgres, example mongo

or something else?

File: scripts/setup/user.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+            "scopes": SCOPE_REGISTRY,
+        },
+    )
+
+    if not response.ok:
+        raise RuntimeError(
+            f"fidesops user permissions creation failed! response.status_code={response.status_code}, response.json()={response.json()}"
+        )
+    else:
+        logger.info(f"Created user with username: {username} and password: {password}")
```
[the remark this answers is not in the document]
Comment (seanpreston) [reply scripts/setup/user.py:70]: I'm dismissing these alerts as they're not constructive in this context — e.g. we need to log the password here so that the script caller can login as the user.

**Comment by ThomasLaPiana** [comment]
@seanpreston looks like a module is missing:

```
Traceback (most recent call last):
  File "/fides/scripts/example_script.py", line 6, in <module>
    from setup.mailchimp_connector import create_mailchimp_connector
  File "/fides/scripts/setup/mailchimp_connector.py", line 7, in <module>
    import setup.secrets as secrets
ModuleNotFoundError: No module named 'setup.secrets'
```

**Comment by TheAndrewJackson** [comment]
@ThomasLaPiana right now the `secrets.py` file is added to the `.gitignore` and not checked in. I think the expectation is people will provide their own secrets file and populate it with everything that is needed. 

I think it was done this way to help prevent accidentally checking secrets into the repo.

**Comment by ThomasLaPiana** [comment]
@TheAndrewJackson in that case, is there documentation about how to do that or what secrets are required? Why use a secrets file instead of the vault like we do for other stuff? I think we can come up with a better solution here, so I'm going to poke at this

**Review by ThomasLaPiana — DISMISSED** [review]
Needs more documentation

**Comment by seanpreston** [comment]
Hey @ThomasLaPiana / @TheAndrewJackson — `secrets.py` is currently checked into our vault, as mentioned in the PR description:

> The required secrets can be found [here](https://start.1password.com/open/i?a=5LPILPRDUBGWBPJWYS6QHHHLRE&v=42yuohsgd3pqfbepwcth7uigue&i=wjk46p25hvcfhaylrb7gycirm4&h=ethyca.1password.com)

Very fair comment on docs — will write some usage instructions out.

**Comment by ThomasLaPiana** [comment]
getting this error, not sure why there's a need to remove the script name from the posargs though?

```
ValueError: list.remove(x): x not in list
nox > Session run_script(load_examples) failed.
```

File: noxfiles/dev_nox.py [hunk 7]
```diff
@@ -44,3 +44,29 @@ def quickstart(session: nox.Session) -> None:
...
+    [
+        nox.param("example_script", id="example_script"),
+    ],
+)
+def run_script(session: nox.Session, script_name: str) -> None:
+    """
+    Run a script from the scripts/ directory. This command will not run
+    the server automatically.
+    """
+    posargs = session._runner.posargs
```
Comment (ThomasLaPiana) [inline noxfiles/dev_nox.py:61]: why is this `session._runner.posargs` instead of `session.posargs`? I just want to make sure I'm not missing anything here, but `session.posargs` is the standard way to access the posargs

File: noxfiles/dev_nox.py [hunk 8]
```diff
@@ -44,3 +44,29 @@ def quickstart(session: nox.Session) -> None:
...
+    posargs.remove(script_name)
```
Comment (ThomasLaPiana) [inline noxfiles/dev_nox.py:62]: this throws an error if it isn't present, is that intentional? I couldn't get it to run like this

File: docs/fides/docs/guides/utility_scripts.md [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+| `setup.postgres_connector.create_postgres_connector` |
+| `setup.mongodb_connector.create_mongodb_connector` |
+| `setup.privacy_request.create_privacy_request` |
+| `setup.subject_identity_verification.verify_subject_identity` |
+
+Each script uses the Fides API to create its corresponding primitive object in Fides config.
+
+## Secrets management
+
+Some third party integrations require secrets to grant access. Custom scripts are configured to read these secrets as constant variables from a `scripts/setup/secrets.py` file. Due to its sensitive nature, this file is not supplied in the repository, and should instead be stored somewhere more secure such as a password manager. It must be a valid Python file and the format of the secrets should be:
```
Comment (ThomasLaPiana) [inline docs/fides/docs/guides/utility_scripts.md:28]: I tried to put a breadcrumb here so that our internal devs would know where the file was likely to be located internally. Should we state it more directly somewhere?

**Comment by ThomasLaPiana** [comment]
hitting this error: 

```
RuntimeError: fides connection test failed! response.status_code=200, response.json()={'msg': 'Secrets updated for ConnectionConfig with key: postgres_connector.', 'test_status': 'failed', 'failure_reason': 'Operational Error connecting to postgres db.'}
nox > Command docker compose run fides python /fides/scripts/load_examples.py failed with exit code 1
nox > Session run_script(load_examples) failed.
```

**Comment by ThomasLaPiana** [comment]
@seanpreston I can't get past this error, can you confirm the script was working for you? All tests are still passing, so I'm thinking something in the script might be incorrect as opposed to an application error? or maybe we've found an edge case

**Comment by NevilleS** [comment]
@ThomasLaPiana this command makes assumptions about what services are running locally. For that postgres connector script to succeed it will need to be running the integration test postgres container. Did you run this alongside with `nox -s dev -- postgres mongodb` so that those are running?

File: noxfiles/dev_nox.py (hunk 7 above)
Comment (seanpreston) [reply noxfiles/dev_nox.py]: Will update — this was the result of me digging around in the interpreter and seeing what's available. The `nox` docs seem quite light on detail here

File: scripts/setup/mailchimp_connector.py [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+logger = logging.getLogger(__name__)
+logging.basicConfig(level=logging.INFO)
+
+
+def create_mailchimp_connector(
+    auth_header: Dict[str, str],
+    key: str = "mailchimp",
+):
+    if not key:
+        key = str(uuid.uuid4())
```
Comment (RobertKeyser) [inline scripts/setup/mailchimp_connector.py:20]: Question about this: 

Am I correct that the only time that the conditional will evaluate to `true` is if the method is called as `create_mailchimp_connector(auth_header=header,key="")`, that is, because we have the default of `mailchimp`. 

If a function explicitly calls this with an empty string, I would expect it to fail instead of generating a uuid for it. Now if the value was omitted, it would then be fine to generate something random, but I would prefer to align the inputs with expected outputs.

Comment (seanpreston) [reply scripts/setup/mailchimp_connector.py:20]: The `not key` check is Falsy, so `key=""` or `key=None` should both trigger it. If the DB object is created without a key, one will be automatically generated for it, so it's best we give it something sensible here. The default value was only meant to be a handy default.

File: docs/fides/docs/guides/utility_scripts.md (hunk 9 above)
Comment (seanpreston) [reply docs/fides/docs/guides/utility_scripts.md:28]: Thanks, I think we don't need to now since the secrets can also be supplied as environment variables

File: noxfiles/dev_nox.py (hunk 7 above)
Comment (seanpreston) [reply noxfiles/dev_nox.py]: Thanks for rectifying this

**Comment by seanpreston** [comment]
Thanks for the thorough reviews @ThomasLaPiana / @NevilleS / @RobertKeyser — I'm going to go ahead and merge this now so these can be put to use in some form. We can add in any precise DX improvements as follow-up tickets.
